### DISCUSSION TITLE
Boolean Data Type

### DISCUSSION THREAD

**Discussion by ghislaineguerin** [mathesar-foundation/mathesar#550] [Maintainer discussions]
The specs related to issue #237  are ready for review. Please take a look at the attached document and follow the [review guidelines](https://wiki.mathesar.org/en/design/process/review-guidelines).

## Document Link
https://wiki.mathesar.org/en/design/specs/boolean-data-type

@mathemancer @kgodey @pavish, your feedback is required by Thursday 19th. Please plan your review so that at least one round of changes can be made and reviewed if needed.

**Comment by kgodey** [comment]
@ghislaineguerin This looks good! Some feedback:

- I think we should consider using system default checkboxes instead of implementing our own, it might be more familiar to the user. Generally, the checkbox icons used don't seem immediately interactive to me.

- I think it would be worth specifying that booleans can also have `NULL` values, and how that would show up in the dropdown. It's also worth explicitly specifying the indeterminate checkbox state.

- I'm wondering if we also want to support other display formats, such as a toggle (like the ones we use for allow null/allow duplicate). We'd have to figure out an indeterminate state for that too.

- The "Custom Values" label might lead users to believe that the actual values are being stored at the database level. This is just a display option and that should be more clear.
  - On a related note, across all data types, we need to make the distinction between database affecting changes and display affecting changes really clear. 
  - It might be useful to have separate sub-sections in each data type menu for "Database Options" and "Display Options". That terminology can be consistent across all data types, and it would make it very clear.

**Comment by ghislaineguerin** [comment]
@pavish Can you please add your feedback regarding the different display options for boolean types? You had some concerns about this when we reviewed it together. Mostly due to not having a clear use case for all the different options (checkbox, dropdown, etc)

**Comment by pavish** [comment]
@kgodey I was wondering if we need multiple display options for booleans, right away. I see the need for dropdown to display different display text instead of true/false (We could also do the same with checkbox, with a label next to it).

I'm wondering how this works with copy/paste into cells and csv imports (when we support appending in the future). Are the boolean values going to be recognised based on this display text?

**Reply by kgodey** [reply]
@pavish I do think we should support checkbox and dropdown. We don't _need_ the toggle, I didn't think it would be very difficult to implement and it seemed like a nice option for users to customize the look of their tables. If it would be difficult to implement, we can skip it for the MVP.

We do plan to improve boolean inference in the future, using the display text could help with that. Once we have user accounts, we could also consider global preferences for how to display booleans.

Could we make it so that copying a boolean cell copies the underlying value (true or false) and not the representation? That would be easiest, I think. That way, we can paste it into any other boolean column and it would be represented accordingly.

**Reply by ghislaineguerin** [reply]
@kgodey It was my thinking as well that copying would result in 'TRUE' or 'FALSE' being added to the clipboard. I think that would work

**Comment by ghislaineguerin** [comment]
@kgodey @pavish I have updated the specs and the data type options to remove the checkbox variations for now. Meaning we'd have both checkbox and dropdown as options but not different checkboxes.

**Reply by kgodey** [reply]
@ghislaineguerin The "boolean type options" prototype link [in the spec document](https://wiki.mathesar.org/en/design/specs/boolean-data-type) links to the prototype for date/time options currently.

**Reply by ghislaineguerin** [reply]
@kgodey thanks, fixed it.

**Reply by kgodey** [reply]
Looks good, thanks!
